include/seastar/rpc/rpc_impl.hh

```diff
@@ -897,9 +897,11 @@ future<> sink_impl<Serializer, Out...>::flush() noexcept {
 template<typename Serializer, typename... Out>
 future<> sink_impl<Serializer, Out...>::close() noexcept {
     return with_semaphore(this->_sem, max_stream_buffers_memory, [this] {
-        // the send and delete queues should be drained already
-        // since we acquired all the semaphore units, so no need to stop them.
+        // break the semaphore to prevent any new messages to be sent
+        this->_sem.broken(stream_closed());
+        return _send_queue.stop().finally([this] {
         return smp::submit_to(this->_con->get_owner_shard(), [this] {
+          return _delete_queue.stop().finally([this] {
```

Comment:
Shouldn't it happen on the original shard as well? Both, send_queue and delete_queue are created and initialized on the same shard, and the process_fut is resolved on the same shard as well.